// styletron-engine-monolithic/dist-browser-cjs/server/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.generateHtmlString = generateHtmlString;

var _injectStylePrefixed = _interopRequireDefault(require("../inject-style-prefixed"));

var _css = require("../css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class StyletronServer {
  constructor(opts = {}) {
    this.opts = opts || {};
    this.cache = {};
  }

  renderStyle(styles) {
    const className = (0, _css.hashCssObject)(styles);

    if (!this.cache[className]) {
      this.cache[className] = (0, _injectStylePrefixed.default)(styles, className, this.opts.prefix || "", Boolean(this.opts.strict)).join("");
    }

    return `${this.opts.prefix || ""}css-${className}`;
  }

  renderFontFace(fontFace) {
    const fontName = (0, _css.hashCssObject)(fontFace);

    if (!this.cache[fontName]) {
      this.cache[fontName] = (0, _css.fontFaceBlockToRule)(`${this.opts.prefix || ""}font-${fontName}`, (0, _css.declarationsToBlock)(fontFace));
    }

    return `${this.opts.prefix || ""}font-${fontName}`;
  }

  renderKeyframes(keyframes) {
    const animationName = (0, _css.hashCssObject)(keyframes);

    if (!this.cache[animationName]) {
      this.cache[animationName] = (0, _css.keyframesBlockToRule)(`${this.opts.prefix || ""}animation-${animationName}`, (0, _css.keyframesToBlock)(keyframes));
    }

    return `${this.opts.prefix || ""}animation-${animationName}`;
  }

  getStylesheets() {
    const hashedNames = [...Object.keys(this.cache)];
    return [{
      css: this.getCss(),
      attrs: {
        "data-hydrate": hashedNames.join(" ")
      }
    }];
  }

  getStylesheetsHtml(className = "_styletron_hydrate_") {
    return generateHtmlString(this.getStylesheets(), className);
  }

  getCss() {
    return [...Object.values(this.cache)].join("");
  }

}

function generateHtmlString(sheets, className) {
  let html = "";

  for (let i = 0; i < sheets.length; i++) {
    const sheet = sheets[i];
    const {
      class: originalClassName,
      ...rest
    } = sheet.attrs;
    const attrs = {
      class: originalClassName ? `${className} ${originalClassName}` : className,
      ...rest
    };
    html += `<style${attrsToString(attrs)}>${sheet.css}</style>`;
  }

  return html;
}

function attrsToString(attrs) {
  let result = "";

  for (const attr in attrs) {
    const value = attrs[attr];

    if (value === true) {
      result += " " + attr;
    } else if (value !== false) {
      result += ` ${attr}="${value}"`;
    }
  }

  return result;
}

var _default = StyletronServer;
exports.default = _default;
